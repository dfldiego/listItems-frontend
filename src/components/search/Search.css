.search_container {
  background-color: var(--yellow);
  display: grid;
  height: 5rem;
  grid-template-columns: repeat(12, 8.33%);
}
/* LOGO */
.search_logo {
  grid-column: 2 / 3;
  grid-row: 1 / 2;
  display: flex;
  align-items: center;
  justify-content: flex-start;
}
.search_logo_img {
  height: 4rem;
  width: 6rem;
}

/* FORM */
.search_form {
  grid-column: 4 / 11;
  grid-row: 1 / 2;
  display: flex;
  align-items: center;
  justify-content: flex-end;
}
@media (min-width: 768px) {
  .search_form {
    grid-column: 3 / 12;
    grid-row: 1 / 2;
    display: flex;
    align-items: center;
    justify-content: flex-end;
  }
}
.search_form_row {
  width: 100%;
  display: flex;
  flex-wrap: nowrap;
}
.input-text {
  width: 100%;
  padding: 0 1rem;
  outline: none;
  border: 1px solid var(--white);
}
/* BOTON SEARCH */
.search_button {
  width: 3.5rem;
  height: 3.5rem;
  background-color: var(--white-moderate);
  border: none;
}
.search_img {
  width: 1.5rem;
  height: 1.5rem;
}
