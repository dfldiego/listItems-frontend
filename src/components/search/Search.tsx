import { useNavigate } from "react-router-dom";

import "./Search.css";
import ic_Search from "../../assets/ic_Search.png";
import Logo_ML from "../../assets/Logo_ML.png";
import useForm from "../../hooks/useForm";
import useQueryString from "../../hooks/useQueryString";

interface FormElements extends HTMLFormControlsCollection {
  searchText: HTMLInputElement;
}
interface SearchTextFormElement extends HTMLFormElement {
  readonly elements: FormElements;
}

type Props = {
  setCurrentPage?: Function;
};

const Search = ({ setCurrentPage }: Props) => {
  const navigate = useNavigate();
  const { q = "" } = useQueryString();

  const { searchText, onInputChange } = useForm({
    searchText: (q as string) || "",
  });

  const handleSubmit = (event: React.FormEvent<SearchTextFormElement>) => {
    event.preventDefault();
    if (searchText.trim().length <= 1) return;
    if (setCurrentPage) setCurrentPage(0);
    navigate(`/items?q=${searchText.toLowerCase().trim()}`);
  };

  const handleClickLogo = () => {
    navigate("/");
  };

  return (
    <div className="search_container" data-testid="search-component-test-id">
      <div className="search_logo">
        <img
          src={Logo_ML}
          alt="MercadoLibre_logo"
          className="search_logo_img"
          data-testid="meli-logo-test-id"
          onClick={handleClickLogo}
        />
      </div>
      <div className="search_form">
        <form
          onSubmit={handleSubmit}
          className="search_form_row"
          aria-label="form"
        >
          <input
            type="text"
            placeholder="Nunca dejes de buscar"
            className="input-text"
            name="searchText"
            value={searchText}
            onChange={({ target }) => onInputChange(target.name, target.value)}
            autoComplete="off"
          />
          <button type="submit" className="search_button">
            <div>
              <img
                src={ic_Search}
                alt="Logotipo Buen Sabor"
                className="search_img"
              />
            </div>
          </button>
        </form>
      </div>
    </div>
  );
};

export default Search;
