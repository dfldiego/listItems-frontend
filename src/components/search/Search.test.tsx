import { fireEvent, render, screen } from "@testing-library/react";
import * as router from "react-router-dom";

import { MockedRouter } from "../../utils/test/mockedRouter";
import Search from "./Search";

jest.mock("query-string", () => ({
  parse: jest.fn(),
}));

// useQueryString (mock custom hook)
jest.mock("../../hooks/useQueryString", () => {
  return () => ({
    q: "algo",
  });
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: jest.fn(),
  useNavigate: jest.fn(),
}));

describe("<Search>", () => {
  const navigate = jest.fn();
  beforeEach(() => {
    jest.spyOn(router, "useNavigate").mockImplementation(() => navigate);
  });

  test("should render correctly", () => {
    const { container } = render(<MockedRouter component={Search} />);
    expect(container).toMatchSnapshot();
  });

  test("should navigate to home page when i click the logo", () => {
    render(<MockedRouter component={Search} />);
    const logo = screen.getByTestId("meli-logo-test-id");
    fireEvent.click(logo);

    expect(navigate).toHaveBeenCalled();
    expect(navigate).toHaveBeenCalledWith("/");
  });

  test("should submit and navigate correctly when the text as more than one letter", () => {
    render(<MockedRouter component={Search} />);

    const form = screen.getByRole("form");
    const input = screen.getByRole("textbox");
    fireEvent.input(input, { target: { value: "zapatillas" } });
    fireEvent.submit(form);

    expect(navigate).toHaveBeenCalled();
    expect(navigate).toHaveBeenCalledWith("/items?q=zapatillas");
  });

  test("should submit correctly but not navigate when the text as one or zero letter", () => {
    render(<MockedRouter component={Search} />);

    const form = screen.getByRole("form");
    const input = screen.getByRole("textbox");
    fireEvent.input(input, { target: { value: "z" } });
    fireEvent.submit(form);

    expect(navigate).not.toHaveBeenCalled();
  });
});
