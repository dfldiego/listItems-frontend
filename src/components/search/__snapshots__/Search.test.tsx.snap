// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search> should render correctly 1`] = `
<div>
  <div
    class="search_container"
    data-testid="search-component-test-id"
  >
    <div
      class="search_logo"
    >
      <img
        alt="MercadoLibre_logo"
        class="search_logo_img"
        data-testid="meli-logo-test-id"
        src="Logo_ML.png"
      />
    </div>
    <div
      class="search_form"
    >
      <form
        aria-label="form"
        class="search_form_row"
      >
        <input
          autocomplete="off"
          class="input-text"
          name="searchText"
          placeholder="Nunca dejes de buscar"
          type="text"
          value="algo"
        />
        <button
          class="search_button"
          type="submit"
        >
          <div>
            <img
              alt="Logotipo Buen Sabor"
              class="search_img"
              src="ic_Search.png"
            />
          </div>
        </button>
      </form>
    </div>
  </div>
</div>
`;
